Fixed #26292, #25945 -- Refactored MigrationLoader.build_graph().

jarekwg opened the pull request:
[Ticket #26292](https://code.djangoproject.com/ticket/26292)

This PR also solves [Ticket #25945](https://code.djangoproject.com/ticket/25945), including the tests from PR #5939 along with @timgraham's recommended changes to them.

@MarkusH, Adding only normal nodes at the first step proved problematic, as replacing migrations can still reference other replacing migrations, as well as there being possible hiccups that could arise with `__first__`. So the solution I found was to add _all_ nodes to the graph (both normal and replacing migrations), including all their listed dependencies, and _then_ iterate over the replacing migrations and either delete them, or the migrations they replace from the graph, remapping the dependencies as this happens. This ordering seems to work well.

Below are some things I would like feedback on:
1. Since we're now creating dummy nodes instead of immediately raising a `NodeNotFoundError`, for `validate_consistency` to give a meaningful error upon finding dangling dummy nodes, it needs to know about the migration that triggered the creation of the given dummy node. To achieve this, I create the exception when I create the dummy node, and then `validate_consistency` raises that exception if that dummy node still exists at the end. However I'm not hugely happy with the implementation (have a look at the code..).
2. Speed? We're creating and deleting more objects now, as opposed to the previous method of resolving everything in a bunch of dicts prior to building the graph. Might this be visibly less performant, or is this not significant enough to be a worry?
3. I haven't seen any tests for `__first__`, `__latest__` or `run_before` being used in conjunction with squashed migrations. Should we add some in this PR?
4. Is there a point to storing both `graph.node_map` and `graph.nodes`? Could the Migration object not stem off the Node object, rather than sitting in a separate dict? Specifically, I don’t like the doubling up between `graph.nodes[x].dependencies` and `graph.node_map[x].parents`. The current implementation updates `Migration.dependencies` to ensure consistency with the graph. My implementation currently doesn't touch `Migration.dependencies` (Hence one of the tests not passing). I’d be inclined to just delete the `Migration.dependencies` (same with `.replaces` and `.run_before`) lists after the graph has been built from them. Or we should just not touch them after they’ve been loaded in (though deleting would prevent confusion and accidental use). Currently the `showmigrations` mcommand uses `Migration.dependencies` rather than `Node.parents`. Should it really be displaying this? Seems a bit weird to me..

Also, `NodeNotFoundError`’s `__repr__` had a bug: have to call `repr(node)` before passing it to `%r` formatter. Fixed.

This is my first refactor of Django code; please pick it to bits with utmost scrutiny. :)

charettes commented on django/db/migrations/exceptions.py:
```diff
@@ -45,15 +45,16 @@ class NodeNotFoundError(LookupError):
-    def __init__(self, message, node):
+    def __init__(self, message, node, triggering_object=None):
         self.message = message
+        self.triggering_object = triggering_object
```
Does `origin` sounds like a better name to you?

jarekwg replied:
Sure does! I'll change it.

MarkusH commented:
First of all, thanks for the PR!

On Tue, Mar 08, 2016 at 05:03:32AM -0800, Jarek Glowacki wrote:
>@MarkusH, Adding only normal nodes at the first step proved problematic, as replacing migrations can still reference other replacing migrations, as well as there being possible hiccups that could arise with `__first__`. So the solution I found was to add _all_ nodes to the graph (both normal and replacing migrations), including all their listed dependencies, and _then_ iterate over the replacing migrations and either delete them, or the migrations they replace from the graph, remapping the dependencies as this happens. This ordering seems to work well.

Yeah, makes sense.

>1. Since we're now creating dummy nodes instead of immediately raising a `NodeNotFoundError`, for `validate_consistency` to give a meaningful error upon finding dangling dummy nodes, it needs to know about the migration that triggered the creation of the given dummy node. To achieve this, I create the exception when I create the dummy node, and then `validate_consistency` raises that exception if that dummy node still exists at the end. However I'm not hugely happy with the implementation (have a look at the code..).

What about having an actual `DummyNode` that holds all information we
need like migration and refering object and in `validate_consistency()`
trigger the real exception?

```python
class DummyNode(Node):
    def __init__(self, migrations, origin):
	# ...
```

>2. Speed? We're creating and deleting more objects now, as opposed to the previous method of resolving everything in a bunch of dicts prior to building the graph. Might this be visibly less performant, or is this not significant enough to be a worry?

If, in the end the what feels like 1000 loops in the `build_graph()` go
away, I don't see too much of a problem here right now.

>3. I haven't seen any tests for `__first__`, `__latest__` or `run_before` being used in conjunction with squashed migrations. Should we add some in this PR?

Feel free to add them as a separate commit if you feel like it.

>4. Is there a point to storing both `graph.node_map` and `graph.nodes`? Could the Migration object not stem off the Node object, rather than sitting in a separate dict? Specifically, I don’t like the doubling up between `graph.nodes[x].dependencies` and `graph.node_map[x].parents`. The current implementation updates `Migration.dependencies` to ensure consistency with the graph. My implementation currently doesn't touch `Migration.dependencies` (Hence one of the tests not passing). I’d be inclined to just delete the `Migration.dependencies` (same with `.replaces` and `.run_before`) lists after the graph has been built from them. Or we should just not touch them after they’ve been loaded in (though deleting would prevent confusion and accidental use). Currently the `showmigrations` mcommand uses `Migration.dependencies` rather than `Node.parents`. Should it really be displaying this? Seems a bit weird to me..

I'd need to double check. The ticket which this is part of is [#24366](https://code.djangoproject.com/ticket/24366). Maybe @knbk can shed some light on it?

>Also, `NodeNotFoundError`’s `__repr__` had a bug: have to call `repr(node)` before passing it to `%r` formatter. Fixed.

Happy to get this merged into 1.9 (and 1.8?) if you provide me with a
separate issue, test and release note.

>This is my first refactor of Django code; please pick it to bits with utmost scrutiny. :)
>You can view, comment on, or merge this pull request online at:

Sure? :stuck_out_tongue_winking_eye: 

>  https://github.com/django/django/pull/6254
>
>-- Commit Summary --
>
>  * Failing test for #25945 -- Referencing squashed migrations
>  * Illustrate dependency graph
>  * Enable insertion of dummy nodes
>  * Slash looks nicer
>  * Catch NodeNotFoundError off validate_consistency
>  * Update comment
>  * Build the graph in steps instead of resolving the whole migration structure first
>  * Use assertEqual for multiline strings to get cleaner error diffs
>  * Fix bug with calling  on
>  * Initially add ALL migrations, then remove non-applicable replacing migrations later
>  * Clean up some comments and remove a redundant dict.
>  * Trailing comma
>
>-- File Changes --
>
>    M django/db/migrations/exceptions.py (5)
>    M django/db/migrations/graph.py (82)
>    M django/db/migrations/loader.py (191)
>    M tests/migrations/test_commands.py (17)
>    M tests/migrations/test_loader.py (67)
>    A tests/migrations/test_migrations_squashed_ref_squashed/__init__.py (0)
>    A tests/migrations/test_migrations_squashed_ref_squashed/app1/1_auto.py (11)
>    A tests/migrations/test_migrations_squashed_ref_squashed/app1/2_auto.py (13)
>    A tests/migrations/test_migrations_squashed_ref_squashed/app1/2_squashed_3.py (18)
>    A tests/migrations/test_migrations_squashed_ref_squashed/app1/3_auto.py (13)
>    A tests/migrations/test_migrations_squashed_ref_squashed/app1/4_auto.py (13)
>    A tests/migrations/test_migrations_squashed_ref_squashed/app1/__init__.py (0)
>    A tests/migrations/test_migrations_squashed_ref_squashed/app2/1_auto.py (13)
>    A tests/migrations/test_migrations_squashed_ref_squashed/app2/1_squashed_2.py (18)
>    A tests/migrations/test_migrations_squashed_ref_squashed/app2/2_auto.py (13)
>    A tests/migrations/test_migrations_squashed_ref_squashed/app2/__init__.py (0)
>
>-- Patch Links --
>
>https://github.com/django/django/pull/6254.patch
>https://github.com/django/django/pull/6254.diff
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/django/django/pull/6254

--

MarkusH commented:
First of all, thanks for the PR!

On Tue, Mar 08, 2016 at 05:03:32AM -0800, Jarek Glowacki wrote:
>@MarkusH, Adding only normal nodes at the first step proved problematic, as replacing migrations can still reference other replacing migrations, as well as there being possible hiccups that could arise with `__first__`. So the solution I found was to add _all_ nodes to the graph (both normal and replacing migrations), including all their listed dependencies, and _then_ iterate over the replacing migrations and either delete them, or the migrations they replace from the graph, remapping the dependencies as this happens. This ordering seems to work well.

Yeah, makes sense.

>1. Since we're now creating dummy nodes instead of immediately raising a `NodeNotFoundError`, for `validate_consistency` to give a meaningful error upon finding dangling dummy nodes, it needs to know about the migration that triggered the creation of the given dummy node. To achieve this, I create the exception when I create the dummy node, and then `validate_consistency` raises that exception if that dummy node still exists at the end. However I'm not hugely happy with the implementation (have a look at the code..).

What about having an actual `DummyNode` that holds all information we
need like migration and refering object and in `validate_consistency()`
trigger the real exception?

```python
class DummyNode(Node):
    def __init__(self, migrations, origin):
	# ...
```

>2. Speed? We're creating and deleting more objects now, as opposed to the previous method of resolving everything in a bunch of dicts prior to building the graph. Might this be visibly less performant, or is this not significant enough to be a worry?

If, in the end the what feels like 1000 loops in the `build_graph()` go
away, I don't see too much of a problem here right now.

>3. I haven't seen any tests for `__first__`, `__latest__` or `run_before` being used in conjunction with squashed migrations. Should we add some in this PR?

Feel free to add them as a separate commit if you feel like it.

>4. Is there a point to storing both `graph.node_map` and `graph.nodes`? Could the Migration object not stem off the Node object, rather than sitting in a separate dict? Specifically, I don’t like the doubling up between `graph.nodes[x].dependencies` and `graph.node_map[x].parents`. The current implementation updates `Migration.dependencies` to ensure consistency with the graph. My implementation currently doesn't touch `Migration.dependencies` (Hence one of the tests not passing). I’d be inclined to just delete the `Migration.dependencies` (same with `.replaces` and `.run_before`) lists after the graph has been built from them. Or we should just not touch them after they’ve been loaded in (though deleting would prevent confusion and accidental use). Currently the `showmigrations` mcommand uses `Migration.dependencies` rather than `Node.parents`. Should it really be displaying this? Seems a bit weird to me..

I'd need to double check. The ticket which this is part of is [#24366](https://code.djangoproject.com/ticket/24366). Maybe @knbk can shed some light on it?

>Also, `NodeNotFoundError`’s `__repr__` had a bug: have to call `repr(node)` before passing it to `%r` formatter. Fixed.

Happy to get this merged into 1.9 (and 1.8?) if you provide me with a
separate issue, test and release note.

>This is my first refactor of Django code; please pick it to bits with utmost scrutiny. :)
>You can view, comment on, or merge this pull request online at:

Sure? :stuck_out_tongue_winking_eye:

jarekwg commented:
> What about having an actual DummyNode that holds all information we
need like migration and refering object and in validate_consistency()
trigger the real exception?

Sounds good. I'll do that.
> Happy to get this merged into 1.9 (and 1.8?) if you provide me with a
separate issue, test and release note.

Will do.
> Sure? :stuck_out_tongue_winking_eye:

Haha, within reason. :)

jarekwg commented:
@MarkusH https://github.com/django/django/pull/6261

jarekwg commented:
Hrm, so I'm not sure whether it isn't more hacky now than it was before.
Using `DummyNode` objects means that when I want to replace them with actual `Node` objects, I'd have to update all of the dependencies that point to the dummy to point to the new node instead..
I chose to just swap out `__class__` instead, so that the dependencies can still reference the same object. It's a bit of a dirty solution though. :/

MarkusH commented on django/db/migrations/graph.py:
```diff
@@ -79,6 +79,29 @@ def descendants(self):
+        """
+        del self.origin
+        del self.error_message
+        self.__class__ = Node
```
This pattern is actually used in multiple places in Django:

```
$ ack --py -C 3 "\.__class__ = "
django/contrib/gis/gdal/field.py
33-        self.ptr = fld_ptr
34-
35-        # Setting the class depending upon the OGR Field Type (OFT)
36:        self.__class__ = OGRFieldTypes[self.type]
37-
38-        # OFTReal with no precision should be an OFTInteger.
39-        if isinstance(self, OFTReal) and self.precision == 0:
40:            self.__class__ = OFTInteger
41-            self._double = True
42-
43-    def __str__(self):

django/contrib/gis/gdal/geometries.py
117-            self.srs = srs
118-
119-        # Setting the class depending upon the OGR Geometry Type
120:        self.__class__ = GEO_CLASSES[self.geom_type.num]
121-
122-    def __del__(self):
123-        "Deletes this Geometry."

django/contrib/gis/geos/geometry.py
112-                6: MultiPolygon,
113-                7: GeometryCollection,
114-            }
115:        self.__class__ = GEOSGeometry._GEOS_CLASSES[self.geom_typeid]
116-
117-        # Setting the coordinate sequence for the geometry (will be None on
118-        # geometries that do not have coordinate sequences)

django/db/models/fields/__init__.py
86-
87-def _empty(of_cls):
88-    new = Empty()
89:    new.__class__ = of_cls
90-    return new
91-
92-
--
487-        # We need to avoid hitting __reduce__, so define this
488-        # slightly weird copy construct.
489-        obj = Empty()
490:        obj.__class__ = self.__class__
491-        obj.__dict__ = self.__dict__.copy()
492-        return obj
493-

django/db/models/sql/query.py
261-        used by clients to update attributes after copying has taken place.
262-        """
263-        obj = Empty()
264:        obj.__class__ = klass or self.__class__
265-        obj.model = self.model
266-        obj.alias_refcount = self.alias_refcount.copy()
267-        obj.alias_map = self.alias_map.copy()

django/utils/tree.py
40-        any extra setting up to do).
41-        """
42-        obj = Node(children, connector, negated)
43:        obj.__class__ = cls
44-        return obj
45-
46-    def __str__(self):
--
55-        Utility method used by copy.deepcopy().
56-        """
57-        obj = Node(connector=self.connector, negated=self.negated)
58:        obj.__class__ = self.__class__
59-        obj.children = copy.deepcopy(self.children, memodict)
60-        return obj
61-

tests/basic/tests.py
376-                return 'did something'
377-
378-        qs = Article.objects.all()
379:        qs.__class__ = CustomQuerySet
380-        qs = qs.none()
381-        with self.assertNumQueries(0):
382-            self.assertEqual(len(qs), 0)
--
653-            def get_queryset(self):
654-                return FakeQuerySet(self.model)
655-        try:
656:            Article._base_manager.__class__ = FakeManager
657-            asos = ArticleSelectOnSave.objects.create(pub_date=datetime.now())
658-            with self.assertNumQueries(3):
659-                asos.save()
--
666-            with self.assertRaises(DatabaseError):
667-                asos.save(update_fields=['pub_date'])
668-        finally:
669:            Article._base_manager.__class__ = orig_class
670-
671-
672-class ModelRefreshTests(TestCase):
```

jarekwg replied:
Oh neat! Guess that works out well then.

MarkusH commented on django/db/migrations/graph.py:
```diff
@@ -108,27 +131,101 @@ def __init__(self):
+            replacement_node = self.node_map[replacement]
+            for child in node.children:
+                child.parents.remove(node)
+                if child.key != replacement:
```
When is that actually a problem? Isn't it only to prevent self-references on replacing migrations?

jarekwg replied:
Yes. To prevent self-referencing.

The replacement node is not aware of the other nodes it is replacing. When there are several nodes for a replacing migration to replace, `build_graph` loops over them, calling `replace_node` several times. After the first node is replaced, the replacement node may become a parent or child of one of the other nodes that it will be replacing. Further `replace_node` calls could then create self-referencing dependencies at the replacement node.

After some thought though, I'm thinking it would be cleaner to not allow the replacement migration to become a parent/child of any of its replacement targets in the first place by turning this function into `replace_nodes` (plural) and giving it all of the replacement targets at once. In this case, before creating new dependencies between the replacement node and some other node, we'd be able to ensure that this other node is not within the set of replacement targets. Do you think this would be cleaner?

MarkusH replied:
Agree, keeping `replace_nodes()` and `remove_replacement_node()` similar sounds like a cleaner approach.

MarkusH replied:
Can you also add a comment about the "self-referencing" issue (or doesn't that apply anymore)?

jarekwg replied:
Will do.

MarkusH commented on django/db/migrations/graph.py:
```diff
@@ -108,27 +131,101 @@ def __init__(self):
+                    child.add_parent(replacement_node)
+            for parent in node.parents:
+                parent.children.remove(node)
+                if parent.key != replacement:
```
Same here...

jarekwg replied:
See explanation above.

MarkusH commented on django/db/migrations/loader.py:
```diff
@@ -174,92 +198,55 @@ def build_graph(self):
+                for replaced in migration.replaces:
+                    self.graph.replace_node(replaced, key)
+            else:
+                # This replacing migration cannot be used. Remove it from
```
"... cannot be used because it is partially applied."

MarkusH commented on django/db/migrations/graph.py:
```diff
@@ -108,27 +131,101 @@ def __init__(self):
+        """
+        self.nodes.pop(key, None)
+        node = self.node_map.pop(key, None)
+        for child in node.children:
```
Why don't you check for `if node` as in `replace_node()`?

jarekwg replied:
Because the replacement node MUST exist for us to reach this point.
In `replace_node()`, the replacement targets may be have been deleted.
Have I missed some special case where this could be a problem though?

MarkusH replied:
Thanks, makes sense. Why not raise a `NodeNotFound` in case the node is not found?

On Mon, Mar 14, 2016 at 10:51:49PM -0700, Jarek Glowacki wrote:
>> +                    child.add_parent(replacement_node)
>> +            for parent in node.parents:
>> +                parent.children.remove(node)
>> +                if parent.key != replacement:
>> +                    replacement_node.add_parent(parent)
>> +                    parent.add_child(replacement_node)
>> +
>> +    def remove_replacement_node(self, key, replaced_nodes):
>> +        """
>> +        Think of this as the inverse operation to `replace_node`. Kind of.
>> +        Removes the replacement node `key` and remaps its child nodes to
>> +        `replaced_nodes` - the list of nodes it would have replaced.
>> +        """
>> +        self.nodes.pop(key, None)
>> +        node = self.node_map.pop(key, None)
>> +        for child in node.children:
>
>Because the replacement node MUST exist for us to reach this point.
>In `replace_node()`, the replacement targets may be have been deleted.
>Have I missed some special case where this could be a problem though?
>
>---
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/django/django/pull/6254/files#r56117909

--

jarekwg replied:
Sure, can do that. I don't think this can happen from within `build_graph()`, but if some other bit of code were to call this function a second time, then there would be an issue. I'll add a unit test for that.

MarkusH replied:
:+1:

MarkusH commented on django/db/migrations/graph.py:
```diff
@@ -108,27 +131,101 @@ def __init__(self):
+                    replaced_node.add_child(child)
+                    child.add_parent(replaced_node)
+        for parent in node.parents:
+            parent.children.remove(node)
```
Where do you add the new parent references?

jarekwg replied:
So in this case I made the assumption that the replacing migration came into existence _after_ its replacement targets. So both the replacing migration and the replacement targets should be properly aware of their ancestry. The only issue (I think) is with future nodes, which might choose to depend on _either_ a replacing migration or one of its replacement targets. So this is the only part I deemed necessary to remap.

Long story short, I assume the parent links are already in place.

Is this a bad assumption? Should we add it just for the sake of completeness, or is there actually a scenario where you think omitting this could cause a problem?

MarkusH replied:
Oh, yeah. That makes sense. Can you leave a comment with an explanation instead of adding that code, please.

MarkusH commented on django/db/migrations/loader.py:
```diff
@@ -174,92 +198,55 @@ def build_graph(self):
+        # Add external dependencies now that the internal ones have been resolved.
+        for key, migration in self.disk_migrations.items():
+            self.add_external_dependencies(key, migration)
+        # Carry out replacements where possible.
```
Can't we already do the validity checking here?

jarekwg replied:
There could be deleted replacement targets with a replacing migration still has them in its `replaces` list. In this case other normal migrations are still allowed to depend on these deleted migrations, but if they do, they'll be creating `DummyNode`s that won't be resolved until after the next section runs.

MarkusH replied:
*gnarf*. Thanks for clarifying.

MarkusH commented:
@jarekwg I looked through the PR. Except for the few comments I left and the broken test case (that should be a minor thing) I'm more than happy :smile: :+1: 

Thanks @darrenfrenkel for the help on the review!

jarekwg commented:
Thanks! I'm more than happy to hear that. :)

I replied to most of your comments (the ones I skipped I'll have fixed in next update).

Regarding the failing test with `showmigrations`, I think that for this PR I'll just add some code to keep the existing expected behaviour (even though I feel it needs improvement) and might open a separate ticket to raise my concerns with it. The problem is not directly related to this refactor, however in order to keep showing the same dependency lists in `showmigrations`, some extra code needs to be added to `build_graph()`, causing unnecessary (albeit very minor) slowdown.

MarkusH commented:
On Mon, Mar 14, 2016 at 11:41:27PM -0700, Jarek Glowacki wrote:
>Thanks! I'm more than happy to hear that. :)

:)

>Regarding the failing test with `showmigrations`, I think that for this PR I'll just add some code to keep the existing expected behaviour (even though I feel it needs improvement) and might open a separate ticket to raise my concerns with it. The problem is not directly related to this refactor, however in order to keep showing the same dependency lists in `showmigrations`, some extra code needs to be added to `build_graph()`, causing unnecessary (albeit very minor) slowdown.

If you feel strongly about the improvements I don't mind to keep them in the PR as a separate commit, considering that it's all about rewriting the loader to make it more sane. At least as long as it makes sense in this PR.

--

jarekwg commented on django/db/migrations/graph.py:
```diff
@@ -207,8 +221,18 @@ def remove_replacement_node(self, replacement, replaced):
+            exc_value.__cause__ = exc
+            if not hasattr(exc, '__traceback__'):
+                exc.__traceback__ = sys.exc_info()[2]
+            six.reraise(NodeNotFoundError, exc_value, sys.exc_info()[2])
```
@MarkusH Having to write all this just to raise an error for a situation that should technically never happen feels awfully verbose. Do you know of a better way?

MarkusH replied:
Looks like you picket that from the `_reraise_missing_dependency()` function in the migration loader. I'm not too firm with Python 2/3 traceback handling. I introduced that pattern in https://github.com/django/django/pull/3280/files#diff-a0eecbfcd2a528c7502db192412c5a16R235 due to the lack of a better suggestion. I'm open for improvements. I certainly looks like too much boilerplate.

jarekwg commented:
Okay, so here's my beef with the dependencies list shown in `showmigrations`:

Currently, it uses the `Migration.dependencies` list.
Currently, this list gets modified in `MigrationLoader.build_graph()` when replacing migrations are applied - the replaced migration is removed from the list and the replacement added.

I don't like this as:
- I don't think the `Migration` object should be getting modified, especially if it's for the sole purpose of preparing a list for when the user MIGHT call the command `showmigrations` with a higher verbosity.
- It's attempting to clean up user input by modifying this `dependencies` list, even though there could still be a lot more mess there in need of cleanup (such as over-specification of dependencies: if migration A depends on both B and C, but migration B also depends on C, then migration A's dependency on C is superfluous but will still get listed).
- It doesn't consider `run_before` constraints, in which case it will not properly explain why the plan has been ordered in a certain way.

I'd prefer to show `node.parents` instead, as this one will always have replacements, `run_before`s and everything else already considered. 
However, there's currently no constraint on `node.parents` to yield only the bare minimum required dependencies, so there may still exist over-specification. To address this we might want a `MigrationGraph.clean_non_critical_dependencies()` function, which would step through every edge in the graph and check if removing it would change the plan. Something like "If I remove A's dependency on C, will C still be listed in A's `ancestors()` list? If so, remove the dependency." Performance-wise I'm not sure whether it would be worth calling `clean_non_critical_dependencies()` after every `build_graph()` (probably wouldn't), but perhaps this function could just be called upon by `showmigrations`, after which `node.parents` would be providing a clean list of only-critical dependencies?

Alternatively we could try to enforce that the graph always has only-critical edges as we add each dependency, though this would likely cause unnecessary slowdown too..

So, the options are:
- Calling `clean_non_critical_dependencies()`, then listing `node.parents` **(this is what I'm going to go with for the time being..)**.
- Enforcing the graph is always clean at the end of every `add_parent` and `add_child` call, then listing `node.parents`.
- Listing `node.parents` without cleaning.
- Listing `Migration` object's unmodified `dependencies` and `run_before` lists as defined in the migration. And maybe the `replaces` list too..
- Listing `Migration` object's `dependencies` and `run_before`, modified by replacements.
- Keeping it as is, listing only `Migration.dependencies` modified by replacements.

As an aside, we might also like to be helpful to the user and inform them of superfluous dependencies they may have defined in their migrations. Though this would be even more work..

Let me know what you think.

MarkusH commented:
> Okay, so here's my beef with the dependencies list shown in `showmigrations`:

We're speaking about the `--plan` option, aren't we?

> Currently, it uses the `Migration.dependencies` list.
> Currently, this list gets modified in `MigrationLoader.build_graph()` when replacing migrations are applied - the replaced migration is removed from the list and the replacement added.
> 
> I don't like this as:

Agreed, at least with regards to your second point.

> - I don't think the `Migration` object should be getting modified, especially if it's for the sole purpose of preparing a list for when the user MIGHT call the command `showmigrations` with a higher verbosity.

:+1: 

> - It's attempting to clean up user input by modifying this `dependencies` list, even though there could still be a lot more mess there in need of cleanup (such as over-specification of dependencies: if migration A depends on both B and C, but migration B also depends on C, then migration A's dependency on C is superfluous but will still get listed).
> - It doesn't consider `run_before` constraints, in which case it will not properly explain why the plan has been ordered in a certain way.

`run_before` should be merged into `dependencies` upon `build_graph()` at which point they do end up in the plan output.

> I'd prefer to show `node.parents` instead, as this one will always have replacements, `run_before`s and everything else already considered.

I suppose the question is what we want the output to be. A pure list of
* A depends on B, C and D; B depends on C
* A depends on B and D; B depends on C

> However, there's currently no constraint on `node.parents` to yield only the bare minimum required dependencies, so there may still exist over-specification. To address this we might want a `MigrationGraph.clean_non_critical_dependencies()` function, which would step through every edge in the graph and check if removing it would change the plan. Something like "If I remove A's dependency on C, will C still be listed in A's `ancestors()` list? If so, remove the dependency." Performance-wise I'm not sure whether it would be worth calling `clean_non_critical_dependencies()` after every `build_graph()` (probably wouldn't), but perhaps this function could just be called upon by `showmigrations`, after which `node.parents` would be providing a clean list of only-critical dependencies?

That's called [Transitive Reduction](https://en.wikipedia.org/wiki/Transitive_reduction)

> Alternatively we could try to enforce that the graph always has only-critical edges as we add each dependency, though this would likely cause unnecessary slowdown too..

Not necessarily. When we manage to keep a transitive reduction of the entire graph from the beginning we could be able to save some time in the `forwards_plan()` and `backwards_plan()` methods. But that's up for benchmarking.

> So, the options are:
> - Calling `clean_non_critical_dependencies()`, then listing `node.parents` **(this is what I'm going to go with for the time being..)**.
> - Enforcing the graph is always clean at the end of every `add_parent` and `add_child` call, then listing `node.parents`.

As mentioned above, those two options probably require some sort of bench mark for small and big projects.

> - Listing `node.parents` without cleaning.

That would probably the most correct though verbose solution wrt the output

> - Listing `Migration` object's unmodified `dependencies` and `run_before` lists as defined in the migration. And maybe the `replaces` list too..

My initial reason for adding `showmigrations --plan` was the ability to figure out why a migration was run before/after another one, i.e. figuring out where dependencies came from. I don't see that with the `node.parents` solution.

> - Listing `Migration` object's `dependencies` and `run_before`, modified by replacements.

That's what should be happening now.

> - Keeping it as is, listing only `Migration.dependencies` modified by replacements.

Not considering `run_python` makes no sense, from my point of view.
 
> As an aside, we might also like to be helpful to the user and inform them of superfluous dependencies they may have defined in their migrations. Though this would be even more work..
> 
> Let me know what you think.

Thanks for the write down. This seems much bigger than what I thought you had in mind and merits its own ticket. Feel free to c&p yours and my comment there.

jarekwg commented:
> We're speaking about the --plan option, aren't we?

Yep, sorry. I should've looked up the name.

>That's called Transitive Reduction

Ah, thanks! I had a feeling this might be a known pattern haha..

> >Listing `Migration` object's `dependencies` and `run_before`, modified by replacements.

>That's what should be happening now.

Ah. So that _is_ the intended implementation. This is the test I saw which initially motivated my disapproval of the current approach because it isn't considering `run_before`. If you're saying it's just a mistake in the test, then that will be an easy fix.


    migrations/test_commands/MigrateTests/test_showmigrations_plan:213

    @override_settings(MIGRATION_MODULES={"migrations": "migrations.test_migrations_run_before"})
    def test_showmigrations_plan(self):
        """
        Tests --plan output of showmigrations command
        """
        ...
        out = six.StringIO()
        call_command("showmigrations", format='plan', stdout=out, verbosity=2)
        self.assertEqual(
            "[ ]  migrations.0001_initial\n"
            "[ ]  migrations.0003_third ... (migrations.0001_initial)\n"
            "[ ]  migrations.0002_second ... (migrations.0001_initial)\n",
            out.getvalue().lower()
        )
        call_command("migrate", "migrations", "0003", verbosity=0)

In that case I don't have as much of an issue with the current _intended_ implementation as I thought I did. I'll open a ticket to discuss it further, especially the transitive reduction stuff, which could turn out to be quite nice.

For the case of this PR, do you think we can just list the 'unreduced' `node.parents`? If done correctly, it should be identical to the current intended output and it'll save us from having to modify `Migration.dependencies`. I've pushed this change up already so you can take a look and confirm. After fixing that one test, all tests pass.

@timgraham: What's the line length for imports? My isort seems to have fallen into a trap with this a few times now. Could we maybe explicitly set it in `setup.cfg`? (I have a default isort.cfg in my root dir which I think might be messing with django's isort's unset line_length value.)

MarkusH commented on django/core/management/commands/showmigrations.py:
```diff
@@ -110,7 +107,7 @@ def print_deps(migration):
                 deps = print_deps(migration)
-            if (migration.app_label, migration.name) in loader.applied_migrations:
-                self.stdout.write("[X]  %s%s" % (migration, deps))
+            if migration.key in loader.applied_migrations:
```
rename `migration` to `node` in the loop?

MarkusH commented on django/core/management/commands/showmigrations.py:
```diff
@@ -91,17 +91,14 @@ def show_plan(self, connection):
                     seen.add(migration)
 
         # Output
         def print_deps(migration):
```
rename `migration` to `node`, please.

MarkusH commented:
> In that case I don't have as much of an issue with the current _intended_ implementation as I thought I did. I'll open a ticket to discuss it further, especially the transitive reduction stuff, which could turn out to be quite nice.

Great!

> For the case of this PR, do you think we can just list the 'unreduced' `node.parents`? If done correctly, it should be identical to the current intended output and it'll save us from having to modify `Migration.dependencies`. I've pushed this change up already so you can take a look and confirm. After fixing that one test, all tests pass.

Sounds correct to me. I would update the help text for the `--plan` flag to include something along the lines "With a verbosity level of 2 or above all direct migration dependencies and reverse dependencies (`run_before`) will be included)"

> @timgraham: What's the line length for imports? My isort seems to have fallen into a trap with this a few times now. Could we maybe explicitly set it in `setup.cfg`? (I have a default isort.cfg in my root dir which I think might be messing with django's isort's unset line_length value.)

jarekwg commented:
Okay, I think I've just about wrapped up everything I wanted to do here.

One final thing I'm not sure about is whether it's possible for a migration to be listed in more than one `replaces` list (ie. replaced by more than one replacing migration). I've made the assumption that it isn't allowed (don't know how it would even make sense if it was), but just thought I'd confirm this, as I'm not sure whether the existing code makes this simplifying assumption.

Please also advise me after your next review as to what extra documentation needs to be provided before a PR like this can be merged in. I'm guessing it's just a changelog entry? Or is there other stuff too?

MarkusH commented:
Migrations must only show up in at most one replaces list. Though we don't enforce it, I think. More interesting is the question of replaced replacing migrations: can a migration that replaces others itself be replaced? That's not working right now, though there's a ticket to add that feature.

jarekwg commented:
Ah, yeah. That would be a nice addition. I think the comment I made on the trac ticket for this PR could enable that. I suggested dumping more information on the migration graph; colouring the nodes (applied/unapplied) and edges (dependency/replacement). Then running some resolver over it that would automatically determine which replacements can/can't be applied and apply/remove them. This would quite easily support nested replacements, by simply resolving the replacements in an intelligent order. I'd be keen to look at that ticket at some point in the future if no one is working on it yet.

auvipy commented:
@MarkusH whats the ticket number if you remember?

MarkusH commented:
> @MarkusH whats the ticket number if you remember?

Which one specifically?

timgraham commented on django/core/management/commands/showmigrations.py:
```diff
@@ -26,7 +26,11 @@ def add_arguments(self, parser):
             '--plan', '-p', action='store_const', dest='format', const='plan',
-            help='Shows all migrations in the order they will be applied.',
+            help=(
+                'Shows all migrations in the order they will be applied. '
```
So, it seems like you changed behavior in addition to simply making refactors. It would be idea if you could split this into several commits to make that more clear and to make review easier.

MarkusH replied:
Yeah, I agree. Taking out this file as a separate commit or PR sounds like a plan.

MarkusH replied:
The `test_commands.py` file would need to be moved along.

MarkusH commented on tests/migrations/test_loader.py:
```diff
@@ -397,3 +397,70 @@ def test_check_consistent_history(self):
+        plan = set(loader.graph.forwards_plan(('app1', '4_auto')))
+        plan = plan - loader.applied_migrations
+        expected_plan = {
+            ('app1', '4_auto'),
```
Could you sort those in reverse order (to be applied first: top ; to be applied last: bottom), please. Same for the other cases in this test.

MarkusH commented on tests/migrations/test_migrations_squashed_ref_squashed/app1/1_auto.py:
```diff
@@ -0,0 +1,11 @@
+class Migration(migrations.Migration):
+
+    operations = [
+        migrations.RunPython(migrations.RunPython.noop)
```
I don't think you actually need operations. This should be sufficient.

```python
from django.db import migrations


class Migration(migrations.Migration):
    pass
```

jarekwg commented:
Rebased.
Moved showmigrations change into separate commit.
Removed superfluous `operations` lists from test migrations.
Fixed order of migrations. There was also an existing test by someone else that had the reverse order (and missing trailing comma!), so I updated that in a separate commit.

MarkusH commented:
buildbot, test this please.

MarkusH commented:
Thanks @jarekwg . Looks good to me. Performance wise there's no really speed-up or slow-down. I guess optimizations as outlined in [#26407](https://code.djangoproject.com/ticket/26407) would be nice but are not too urgent.

MarkusH commented:
Thanks for your work. Merged in 509379a161d179f9e973fbaf6393e879531756f6, c8df17b6128a5311d9375da2b4a4885337b60e42, 80bf3135d08d7c050cfa603a8645096ee5d18534 🍪
